src/utils/LoopStateMachine/LoopStateMachine.jl

```diff
@@ -0,0 +1,101 @@
+"""
+    LoopStateMachine
+
+Utilities for returning state from functions that run inside a loop.
+
+This is used in e.g clipping, where we may need to break or transition states.
+
+The main entry point is to return an [`Action`](@ref) from a function that 
+is wrapped in a `@controlflow f(...)` macro in a loop.  When a known `Action`
+(currently, `Continue`, `Break`, `Return`, or `FullReturn`) is returned, it is processed
+by the `@controlflow` macro, which allows the function to break out of the loop
+early, continue to the next iteration, or return a value, without being 
+syntactically inside the loop.
+"""
+module LoopStateMachine
+
+struct Action{name, T}
```

Comment:
Why not make `Action` an abstract type with the options as concrete structs rather than consts? its simpler and shows more reliably in the REPL

Reply:
Thinking about this it may be faster to use enums?